### DISCUSSION TITLE
MQTT setup

### DISCUSSION THREAD

**Discussion by dystechnic** [Daveiano/weewx-wdc#203] [Q&A] [answered]
I'm trying to get real-time info through MQTT but I keep getting "_Could not connect to live weather station._" on the page.
I have a working MQTT server with weewx-mqtt installed on the same host as weewx. I can see all values of the instruments in the loop topic, so Weewx send information to the MQTT server.

This is what I have in my skin.conf:

    [[mqtt]]
        mqtt_websockets_enabled = 1
        mqtt_websockets_host = 127.0.0.1
        mqtt_websockets_port = 9001
        mqtt_websockets_ssl = 0
        mqtt_websockets_topic = weather/loop

And these sections are in weewx.conf

    [[MQTT]]
        server_url = mqtt://127.0.0.1:1883
        topic = weather
        unit_system = METRIC
        binding = loop, archive
        log_success = False

     [[WdcReport]]
        skin = weewx-wdc
        enable = true
        lang = nl
        [[[Extras]]]
            [[[[mqtt]]]]
                mqtt_websockets_enabled = 1
                mqtt_websockets_host = 127.0.0.1
                mqtt_websockets_port = 9001
                mqtt_websockets_ssl = 0
                mqtt_websockets_topic = weather/loop

To be complete, this is my MQTT config file:

```
persistence false
#MQTT
listener 1883
protocol mqtt

#websockets
listener 9001
protocol websockets

allow_anonymous true
```

But I keep getting the mentioned warnong on the page.
Any Idea what I am missing?

**Comment by Daveiano** [comment]
Sorry for the late reply, was quite busy the last days...

So I am not an expert in MQTT, but here are some possible starting points for investigation:

**1.** On the machine where WeeWX is running, try to connect to your MQTT server via `mosquitto_sub`. Eg. on Ubuntu (see https://stackoverflow.com/a/57715775):

```
sudo apt install mosquitto-clients
mosquitto_sub -h test.mosquitto.org -t "#" -v
```

You should see the submitted messages in real time?

**2.** Check your MQTT Logs: Something like `sudo tail -f /var/log/mosquitto/mosquitto.log` depending on your log file location.

When you open the WeeWX site it should display something like: 

`New client connected from ::ffff:79.252.85.15:59334 as wewx-wdc-website-715180435 (p2, c1, k60).`

**3.** In the Browser open the Console (via F12 for most browsers, more info here https://screenful.com/guide/how-to/how-to-open-the-browser-developer-console). In the console you should see an error message?

![Screenshot from 2023-10-11 00-48-46](https://github.com/Daveiano/weewx-wdc/assets/3184897/d9603442-edcb-43bf-b385-b3a21584e9d6)

**Reply by dystechnic** [reply]
Hi.

I can understand that you are busy. The wait is not a problem.

`mosquitto_sub -h test.mosquitto.org -t "#" -v` leads to a realtime output of the sensor data. I can also connect to mqtt via a remote machine and see the data there. So mqtt is working and weewx is putting data on the bus.

The output of `sudo tail -f /var/log/mosquitto/mosquitto.log` is:
```
1697092980: mosquitto version 2.0.11 starting
1697092980: Config loaded from /etc/mosquitto/mosquitto.conf.
1697092980: Opening ipv4 listen socket on port 1883.
1697092980: Opening ipv6 listen socket on port 1883.
1697092980: Opening websockets listen socket on port 9001.
1697092980: mosquitto version 2.0.11 running
1697092981: New connection from 127.0.0.1:60739 on port 1883.
1697092981: New client connected from 127.0.0.1:60739 as weewx_9e9da3cf (p2, c1, k60).
```
MQTT is listening on 1883 and the websocket port is 9001. Weewx connects tot the mqtt server.
Nmap on the host shows the same ports as open.

In the browser (Chromium) the console shows some errors:

```
WebSocket connection to 'ws://127.0.0.1:9001/mqtt' failed: 
v._doConnect @ live-updates.js:1

WebSocket connection to 'ws://127.0.0.1:9001/mqtt' failed: 
v._doConnect @ live-updates.js:1

MQTT WS: onConnect Failure.
 
Uncaught Error: AMQJS0011E Invalid state not connected.
    at v.subscribe (live-updates.js:1:16257)
    at Client.subscribe (live-updates.js:1:36078)
    at Object.onFailure (live-updates.js:1:45418)
    at v._disconnected (live-updates.js:1:29072)
    at v._on_socket_error (live-updates.js:1:26272)
    at WebSocket.<anonymous> (live-updates.js:1:7020)

```

**Comment by dystechnic** [comment, marked as answer]
I've made some progress today.
If I use the assigned IP address (192.168.x.x) in the MQTT config part of [[WdcReport]] in weewx.conf and in skin.conf  it started working. When I use localhost or 127.0.0.1 in these config parts it does not work.
In the [[MQTT]] part in weewx.conf it does not matter what I use there. assigned IP, localhost or 127.0.0.1 itall works.

For now I use the assigned IP address but that makes the config file less portable between my test and operational system.

Regards.

**Reply by Daveiano** [reply]
Ok while I am happy that you could solve this, I have no idea why it solves it. Shouldn't make any difference?

**Reply by dystechnic** [reply]
No, it should not. Localhost could be a possible issue because that default to IPv6 ::1 and I don't know if IPv6 and Mosquitto go well together. But 127.0.0.1 and 127.0.1.1 should work.

**Reply by gary-hammer** [reply]
Old topic but for others who have this issue.
The reason localhost and 127.0.0.1 do not work as the MQTT server value for the skin is that all of these values indicate the local machine. On the machine weewx runs on, it is the correct address. On any other machine, it says "Look for this here" and of course it is not "here" it is on the weewx machine.
